draft-ietf-quic-qlog-quic-events.md

```diff
@@ -635,10 +635,37 @@ Most of these settings are typically set once and never change. However, they
 are usually set at different times during the connection, so there will
 regularly be several instances of this event with different fields set.
 
-Note that some settings have two variations (one set locally, one requested by the
-remote peer). This is reflected in the `owner` field. As such, this field MUST be
-correct for all settings included a single event instance. If you need to log
-settings from two sides, you MUST emit two separate event instances.
+Some parameters have two variations: one set locally, and one requested by the
+remote peer, and both are valid and used during the connection. This is
+reflected in the `owner` field. As such, this field MUST be correct for all
+settings included a single event instance. If you need to log settings from two
+sides, you MUST emit two separate event instances.
+
+In contrast, some parameters are the result of negotiation between peers. In
+this case, one endpoint generally suggests a possible value for a parameter and
+its peer then suggests another possible value. The final value used by both
+endpoints is the result of some parameter-specific function applied over the two
+suggested values. Such negotiation SHOULD be expressed as a series of
+`parameters_set` events with the appropriate `owner` field value set. For
+example:
+
+- The local endpoint logs a `parameters_set` event with an `owner` of `local`
+  reflecting the parameter value(s) it would like to use.
+- The local endpoint receives parameter values from the peer and logs a
+  `parameters_set` event with an `owner` of `remote` reflecting the desired
+  value(s) expressed by the peer.
+- The local endpoint logs a `parameters_set` event with an `owner` of `local`
+  reflecting the resulting parameter value(s) determined from the suggested
+  parameter value(s) communicated by the local and remote endpoints.
+
+Consequently, each parameter value in a `parameters_set` event supersedes the
+previous value of that parameter if present. If a parameter does not appear in a
+given `parameters_set` event, its value is unchanged from the previous event
+which set that parameter. If a parameter never appears in a `parameters_set`
```

Comment:
This doesn't work for 0-RTT, does it?

Reply:
0-RTT is special cased in the paragraph below this one and should use `parameters_restored`. (That's not to say that that is 100% watertight, I might have missed something in that flow, but it's indeed not the intent to use only `parameters_set` for 0-RTT and this new text is not intended for 0-RTT indeed).

Reply:
I don't think this event should be logged more than once per connection. Transport parameters are only exchanged once, after all. If we have a separate owner value as suggested above, this property would also hold in qlog.

Reply:
This event will be logged at least twice per connection by definition, since there are two sets of TPs: one for the local entity and one for the peer. This event also tracks more state than just transport parameters (e.g., TLS level params like resumption or early data) that are probably set in a different codepath/at a different time and benefit from having multiple `parameters_set`.

I agree with the subset of your statement that the event probably shouldn't be logged more than once per connection for specific fields with the same owner though. 

Still, for me personally the most important part here is `If a parameter does not appear in a given parameters_set event, its value is unchanged from the previous event which set that parameter.` to help make clear that subsequent `parameters_set` do not invalidate earlier values, which I agree with @hlandau is not 100% clear right now.